### PR TITLE
Rewrite from AsyncTask to Kotlin coroutines

### PR DESCRIPTION
**Pull request by M1ST4KE**
# Why

`AsyncTask` is deprecated since api level 30, it's recommended to use the standard java.util.concurrent or Kotlin concurrency utilities instead.

# How

Rewrote classes extending AsyncTask, to use kotlin concurrency utilities

# Test Plan

Checked module in bare-expo app
Ran test-suite tests

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt [hunk 1]
```diff
@@ -158,41 +169,47 @@ internal class BarCodeScannerViewFinder(
...
+          )
+          if (result != null) {
+            withContext(Dispatchers.Main) {
+              launch {
+                barCodeScannerView.onBarCodeScanned(result)
+              }
+            }
+//              Handler(Looper.getMainLooper()).post {
+//                barCodeScannerView.onBarCodeScanned(result)
+//              }
```
Comment (lukmccall) [inline packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt:206]:
```suggestion
```
?

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt [hunk 2]
```diff
@@ -35,6 +40,7 @@ internal class BarCodeScannerViewFinder(
   private var isChanging = false
   private var camera: Camera? = null
 
+
```
Comment (lukmccall) [inline packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt:43]:
```suggestion
```
?

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt [hunk 3]
```diff
@@ -4,23 +4,28 @@ import android.content.Context
...
 
 internal class BarCodeScannerViewFinder(
-  context: Context,
-  private var cameraType: Int,
-  private var barCodeScannerView: BarCodeScannerView,
-  private val moduleRegistryDelegate: ModuleRegistryDelegate
+    context: Context,
+    private var cameraType: Int,
+    private var barCodeScannerView: BarCodeScannerView,
+    private val moduleRegistryDelegate: ModuleRegistryDelegate
```
Comment (lukmccall) [inline packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt:25]: ?

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt [hunk 4]
```diff
@@ -101,11 +112,11 @@ internal class BarCodeScannerViewFinder(
...
-            ExpoBarCodeScanner.instance.getBestSize(
-              temporaryParameters.supportedPictureSizes,
-              Int.MAX_VALUE,
-              Int.MAX_VALUE
-            )
+              ExpoBarCodeScanner.instance.getBestSize(
+                  temporaryParameters.supportedPictureSizes,
+                  Int.MAX_VALUE,
+                  Int.MAX_VALUE
+              )
```
Comment (lukmccall) [inline packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt:119]: ?

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt [hunk 5]
```diff
@@ -158,41 +169,47 @@ internal class BarCodeScannerViewFinder(
...
-  private inner class BarCodeScannerAsyncTask(private val camera: Camera?, private val mImageData: ByteArray, barCodeScannerView: BarCodeScannerView) : AsyncTask<Void?, Void?, Void?>() {
-    init {
-      this@BarCodeScannerViewFinder.barCodeScannerView = barCodeScannerView
-    }
-
-    override fun doInBackground(vararg params: Void?): Void? {
-      if (isCancelled) {
-        return null
-      }
+  private fun barCodeScannerAsyncTask(camera: Camera?, mImageData: ByteArray) {
```
Comment (bbarthec) [inline packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewFinder.kt:180]: I dislike the idea of using `*AsyncTask` sufix in the method name, while it has nothing with it anymore.
```suggestion
  private fun scanForBarcodes(camera: Camera?, mImageData: ByteArray) {
```

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewManager.kt [hunk 6]
```diff
@@ -1,11 +1,15 @@
 package expo.modules.barcodescanner
 
 import android.content.Context
+import android.util.Log
```
Comment (bbarthec) [inline packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewManager.kt:4]: Unused import? 🤔 
```suggestion
```

File: packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewManager.kt [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
 import expo.modules.core.ModuleRegistry
 import expo.modules.core.ModuleRegistryDelegate
 import expo.modules.core.ViewManager
 import expo.modules.core.interfaces.ExpoProp
 import expo.modules.interfaces.barcodescanner.BarCodeScannerSettings
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.cancel
```
Comment (bbarthec) [inline packages/expo-barcode-scanner/android/src/main/java/expo/modules/barcodescanner/BarCodeScannerViewManager.kt:12]: Unused imports? 🤔 
```suggestion
```

File: packages/expo-sharing/CHANGELOG.md [hunk 8]
```diff
@@ -12,6 +12,7 @@
 
 - Migrated from `@unimodules/core` to `expo-modules-core`. ([#13757](https://github.com/expo/expo/pull/13757) by [@tsapeta](https://github.com/tsapeta))
 - Rewrote Android part from Java to Kotlin ([#14010](https://github.com/expo/expo/pull/14010) by [@m1st4ke](https://github.com/m1st4ke))
+- Migrated from `AsyncTask` to Kotlin concurrency utilities. ([#14029](https://github.com/expo/expo/pull/14029) by [@m1st4ke](https://github.com/m1st4ke))
```
Comment (bbarthec) [inline packages/expo-sharing/CHANGELOG.md:15]: Why not to name them explicitly? 😅 
```suggestion
- Migrated from `AsyncTask` to Kotlin coroutines. ([#14029](https://github.com/expo/expo/pull/14029) by [@m1st4ke](https://github.com/m1st4ke))
```

**Review by bbarthec — CHANGES_REQUESTED** [review]
I've got a few remarks, but overall the code looks solid 👍

File: packages/expo-barcode-scanner/android/build.gradle [hunk 9]
```diff
@@ -84,4 +84,6 @@ dependencies {
   api 'com.google.zxing:core:3.3.3'
 
   implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:${safeExtGet('kotlinVersion', '1.4.21')}"
+  implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core:1.4.3")
+  implementation("org.jetbrains.kotlinx:kotlinx-coroutines-android:1.4.1")
```
[the remark this answers is not in the document]
Comment (M1ST4KE) [reply packages/expo-barcode-scanner/android/build.gradle:88]: Latest coroutines version requires latest kotlin version - `1.5.20`, we are using `1.4.21`

**Review by bbarthec — APPROVED** [review]
LGTM 👍

**Pull request merged**